llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -18988,7 +18984,7 @@ SDValue DAGCombiner::visitFP_ROUND(SDNode *N) {
     // single-step fp_round we want to fold to.
     // In other words, double rounding isn't the same as rounding.
     // Also, this is a value preserving truncation iff both fp_round's are.
-    if (DAG.getTarget().Options.UnsafeFPMath || N0IsTrunc)
+    if (N->getFlags().hasAllowContract() || N0IsTrunc)
```

Comment:
This is the only one I'm iffy about - can you split this one into a separate patch?